### PR TITLE
Deal better with restarted peers

### PR DESCRIPTION
**Pull request by bboreham**
Improves the behaviour reported at #28.

### CODE REVIEW

File: local_peer.go [hunk 1]
```diff
@@ -180,7 +180,10 @@ func (peer *localPeer) handleAddConnection(conn Connection) error {
 	}
 	_, isConnectedPeer := peer.router.Routes.Unicast(toName)
 	peer.addConnection(conn)
-	if isConnectedPeer {
+	if isRestartedPeer {
+		conn.log("connection added (restarted peer)")
+		peer.router.sendAllGossipDown(conn)
+	} else if isConnectedPeer {
```
Comment (rade) [inline local_peer.go]: use a `switch`.

**Comment by rade** [comment]
I reckon the `isConnectedPeer` should be lifted out, i.e. the flag passed to `doAddConnection` should simply be an indication as to whether this is a "new" peer, where "new" includes "restarted".

That indicator should be calculated by `registerRemote`: a peer is "new" if a) it is not present in the topology, or b) is present and has a different UID, or c) is present and has no connections.

**Comment by rade** [comment]
> c) is present and has no connections.

The reason we need this is because we may, temporarily, have multiple connections to the same peer.

Looking at this again, perhaps we should leave the code in this PR as is. After all, it does address the problem, and is quite straightforward.

**Comment by bboreham** [comment]
I have rebased and made the change suggested at https://github.com/weaveworks/mesh/pull/29#issuecomment-199783229.

A couple of aspects are now potentially hard to follow: the computation of `isJustAddedPeer` which relies on the behaviour of `fetchWithDefault`, and the absence of locking round `len(conn.remote.connections)`, which (I think) relies on `remote` not being the local peer and `connections` never being modified after creation on a remote peer.

CircleCI tests have failed, which seems to be in the `etcd` code and unrelated.

**Comment by rade** [comment]
As discussed, the "has no connections" tests is probably too narrow. When two clusters come together, a 2nd in-progress connection between them may well find the other peer in the topology, with lots of connections to peers on its side of the cluster. We'd still need to send it the full gossip, so it can learn about *our* side of the cluster.

The symmetric unicast route check of the previous version is a conservative approximation of whether another peer is part of our cluster, which is kinda what we want here.

**Comment by bboreham** [comment]
I removed the last commit, so now we are back to checking the unicast topology.

**Comment by rade** [comment]
LGTM. @peterbourgon what's up with the CI failures?

**Comment by peterbourgon** [comment]
@rade It's the vendoring nonsense. Needs a rebase on master to get [this](https://github.com/weaveworks/mesh/commit/8aabe084b60a9abbc442116d9f49f1ede5edb9d0).

**Comment by peterbourgon** [comment]
I'll do that...
